internal/schema/schema.go

```diff
@@ -0,0 +1,343 @@
+package schema
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/cerbos/cerbos/internal/observability/tracing"
+	"strings"
+	"sync"
+
+	"github.com/qri-io/jsonschema"
+	"go.uber.org/zap"
+	"google.golang.org/protobuf/types/known/structpb"
+
+	enginev1 "github.com/cerbos/cerbos/api/genpb/cerbos/engine/v1"
+	schemav1 "github.com/cerbos/cerbos/api/genpb/cerbos/schema/v1"
+	"github.com/cerbos/cerbos/internal/config"
+	"github.com/cerbos/cerbos/internal/observability/logging"
+	"github.com/cerbos/cerbos/internal/storage"
+)
+
+type Manager struct {
+	conf                *Conf
+	log                 *zap.SugaredLogger
+	store               storage.Store
+	mu                  sync.RWMutex
+	principalProperties map[string]string
+	resourceProperties  map[string]map[string]string
+	schema              *schemav1.Schema
+	principalSchema     *jsonschema.Schema
+	resourceSchemas     map[string]*jsonschema.Schema
+}
+
+func New(ctx context.Context, store storage.Store) (*Manager, error) {
+	conf := &Conf{}
+	if err := config.GetSection(conf); err != nil {
+		return nil, fmt.Errorf("failed to get config section %q: %w", confKey, err)
+	}
+
+	mgr := &Manager{
+		conf:                conf,
+		log:                 zap.S().Named("schema"),
+		store:               store,
+		principalProperties: make(map[string]string),
+		resourceProperties:  make(map[string]map[string]string),
+	}
+
+	if err := mgr.ReadOrUpdateSchemaFromStore(ctx); err != nil {
+		if conf.Enforcement == EnforcementReject {
+			return nil, fmt.Errorf("schema file not found: %w", err)
+		} else {
+			mgr.log.Warnf("schema file not found: %s", err)
+		}
+	}
+
+	store.Subscribe(mgr)
+
+	return mgr, nil
+}
+
+func (m *Manager) Validate(ctx context.Context, input *enginev1.CheckInput) error {
+	ctx, span := tracing.StartSpan(ctx, "schema.Validate")
+	defer span.End()
+	m.mu.RLock()
+	defer m.mu.RUnlock()
```

Comment:
There are a couple of problems here:
- The lock is held for longer than necessary (we only need the lock to check whether the schema is nil or not)
- There's nested locking because `validatePrincipal` and other methods called from this function obtain the lock again

I would just use the lock once to grab copies of the protected fields and then pass those copies around to the other methods. It's better to avoid locking as much as possible.

```suggestion
	m.mu.RLock()
	schema := m.schema
	principalSchema := m.principalSchema
	// whatever other protected fields you need in the body below
	m.mu.RUnlock()
```